Tasks/SonarQubePostTest/PRCA/PostComments-Module.psm1

```diff
@@ -181,39 +299,190 @@ function GetPullRequestId
     
-    $newDiscussionThreads = CreateDiscussionThreads $comments
+    $newDiscussionThreads = CreateDiscussionThreads $messages $messageSource
     PostDiscussionThreads $newDiscussionThreads 
 }
 
+# region Filter Comments
+function FilterComments 
+{
+    param ([Array]$messages, 
+    [ValidateNotNullOrEmpty()][string]$messageSource, 
+    [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionThread]]$existingThreads,
+    [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionComment]]$existingComments)
+    
+    Write-Verbose "Filtering comments before posting"
+    
+    $messages = FilterCommentsByPath $messages
+    $messages = FilterPreExistingComments $messages $messageSource $existingThreads $existingComments
+    $messages = FilterMessagesByNumber $messages
+    
+    return $messages
+}
+
+function FilterCommentsByPath
+{
+    param ([Array]$messages)
+    
+    $modifiedFilesInPr = GetModifiedFilesInPR
+    Write-Verbose "Files changed in this PR: $modifiedFilesInPr"
+    
+    $countBefore = $messages.Count
+    $messages = $messages | Where-Object {$modifiedFilesInPr.Contains($_.RelativePath)}
+    $commentsFiltered = $countBefore - $messages.Count 
+    
+    Write-Host "$commentsFiltered message(s) were filtered because they do not belong to files that were changed in this PR"
+    
+    return $messages
+}
+
+#
+# A message is said to be pre-existing if:
+# 1. a comment with the same content exists 
+# 2. and it belongs to the same file 
+# 3. and the thread was created by the same logic, i.e. the same commentSource 
+#
+# Remark: comments move arround so Line cannot be used. 
+#
+function FilterPreExistingComments
+{
+     param ([Array]$messages, 
+     [ValidateNotNullOrEmpty()][string]$messageSource, 
+     [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionThread]]$existingThreads,
+     [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionComment]]$existingComments)
+     
+     $sw = new-object "Diagnostics.Stopwatch"
+     $sw.Start();
+     
+     $countBefore = $messages.Count
+     $messages = $messages | Where-Object { !(MessageHasMatchingComments $_)}
+     $commentsFiltered = $countBefore - $messages.Count 
+     
+     Write-Host "$commentsFiltered message(s) were filtered because they were already present"
+     Write-Verbose "Filtering out $($existingComments.Count) existing comments took $($sw.ElapsedMilliseconds) ms"
+     
+     return $messages
+}
+
+function MessageHasMatchingComments
+{
+    param ([ValidateNotNull()][PSObject]$message)
+    
+    $matchingComments = GetMatchingComments $_ $messageSource $existingThreads $existingComments
+    
+    return ($matchingComments.Count -gt 0)
+}
+
+#TODO: can be optimized by using a map of <Thread,List<Comments>> instead of 2 flat lists
```

Comment:
Can be a bit optimized by that, since we create 1 thread for each comment - I do not expect a massive gain